src/components/drag-and-drop-list/DragAndDropList.tsx

```diff
@@ -0,0 +1,137 @@
+import {useRef, useState} from "react";
+import "./drag-and-drop-list.css";
+import {DragAndDropTodoItem} from "../drag-and-drop-todo-item/DragAndDropTodoItem";
+import {GhostItem} from "../ghost-item/GhostItem";
+
+interface DragAndDropListProps {
+  title: string
+  listItems: Array<string>
+  listId: number
+  setListItems: (listItems: Array<string>) => void;
+  onItemDropped?: (item: string, sourceListId: number, targetListId: number) => void;
+}
+interface Coordinates {
+  x: number
+  y: number
+}
+
+export function DragAndDropList(props: DragAndDropListProps) {
+  const { title, listItems: items, setListItems: setItems, listId, onItemDropped } = props;
+
+  const [dragStartIdx, setDragStartIdx] = useState<number | null>(null);
+  const [isDragging, setIsDragging] = useState(false);
+  const [ghostPosition, setGhostPosition] = useState<Coordinates>({ x: 0, y: 0 });
+  const [draggedItem, setDraggedItem] = useState<string | null>(null);
+  const [hoverIndex, setHoverIndex] = useState(null);
+  const [isOverDraggableArea, setIsOverDraggableArea] = useState(false);
+
+  const droppableRef = useRef(null);
+
+  function handleDragStart(e, id: string) {
+    console.log(+id)
+    setDragStartIdx(+id)
+    setDraggedItem(items[+id]);
+    setIsDragging(true);
+    setGhostPosition({ x: e.clientX, y: e.clientY });
+    e.dataTransfer.setData('text/plain', JSON.stringify({id, name: items[+id], listId}));
+
+    const transparentImage = new Image();
+    transparentImage.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"; // 1x1 transparent GIF
+    e.dataTransfer.setDragImage(transparentImage, 0, 0);
+  };
+
+  function handleDrag(e) {
+    setGhostPosition({ x: e.clientX, y: e.clientY });
+  };
+
+  function handleDragEnd() {
+    setIsDragging(false);
+    setDraggedItem(null);
+  };
+
+  function handleDragOver(e, index) {
+    e.preventDefault();
+    setHoverIndex(index);
+    setIsOverDraggableArea(true);
+  };
+
+  function handleDrop(e) {
+    const json = e.dataTransfer.getData('text/plain')
+    const {id, name, listId: draggingItemsListId} = JSON.parse(json)
+
+    if(draggingItemsListId === listId) {
+      handleDropOnSameList(name, id)
+    } else {
+      handleDropOnDifferentList(name, +draggingItemsListId)
+    }
+  };
+
+  function handleDropOnDifferentList(name: string, sourceId: number) {
+    items.splice(hoverIndex, 0, name);
+    setItems([...items])
+    setIsDragging(false);
+    setHoverIndex(null);
+
+    if(onItemDropped) {
+      if(sourceId === 1) onItemDropped(name, 1, 2);
+      else onItemDropped(name, 2, 1);
+    }
+  }
+
+  function handleDropOnSameList(name: string, id: string) {
+    if(dragStartIdx < hoverIndex) {
+      items.splice(+id, 1);
+      items.splice(hoverIndex - 1, 0, name);
+    } else {
+      items.splice(+id, 1);
+      items.splice(hoverIndex, 0, name);
+    }
+    setItems(items)
```

Comment:
Again, you set the state without setState function and this can cause you some troubles. I would do this instead.
```suggestion
    const updatedItems = [...items];
    updatedItems.splice(+id, 1);
    
    if (dragStartIdx < hoverIndex) {
      updatedItems.splice(hoverIndex - 1, 0, name);
    } else {
      updatedItems.splice(hoverIndex, 0, name);
    }
    
    setItems(updatedItems);
```

Reply:
Or even better
```suggestion
    const updatedItems = items.filter((item) => item !== id);
    const targetIndex = dragStartIdx < hoverIndex ? hoverIndex - 1 : hoverIndex;
    
    setItems(updatedItems.toSpliced(targetIndex, 0, name));
```